client/lib/perfmon/index.js

```diff
@@ -0,0 +1,173 @@
+// to record new placeholders and remove nodes that are no longer placeholders)
+function observeDomChanges( MutationObserver ) {
+	// if anything scrolls, check if any of our placeholder elements are in view,
+	// but not more than a few times a second
+	window.addEventListener('scroll', debounce(checkForVisiblePlaceholders.bind(null, 'scroll'), 200), true);
+
+	// if the user navigates, stop the current event and proceed to the next one
+	page( function( context, next ) {
+		// send "placeholder-wait-navigated" event
+		checkForVisiblePlaceholders( 'navigate' );
+		next();
+	} );
+
+	// this is fired for matching mutations (childList and class attr changes)
+	var observer = new MutationObserver(function(mutations, observer) {
+		
+		// record all the nodes that match our placeholder classes in the "activePlaceholders" array
+		mutations.forEach( recordPlaceholders );
+
+		// remove any nodes from activePlaceholders that are no longer placeholders
+		// check each node for:
+		// a. whether it's still in the DOM at all, and if so:
+		// b. whether it still has a placeholder class
+		var removed = remove( activePlaceholders, function( node ) {
+			return !OBSERVE_ROOT.contains( node ) || !isPlaceholder( node );
+		} );
+
+		checkForVisiblePlaceholders( 'mutation' );
+
+	} );
+
+	observer.observe( OBSERVE_ROOT, {
+	  subtree: true,
+	  attributes: true,
+	  childList: true,
+	  attributeFilter: ['class']
+	} );
+}
+
+// check if there are any placeholders on the screen,
+// and trigger a timing event when all the placeholders are gone or the user
+// has navigated
+function checkForVisiblePlaceholders( trigger ) {
+	// determine how many placeholders are active in the viewport
+	const visibleCount = activePlaceholders.reduce( 
+		function( count, node ) { 
+			return count + ( isElementVisibleInViewport( node ) ? 1 : 0 ); 
+		}, 0 
+	);
+
+	// record event and reset timer if all placeholders are loaded OR user has just navigated
+	if ( placeholdersVisibleStart && ( visibleCount === 0 || trigger === 'navigate' ) ) {
+		// tell tracks to record duration
+		var duration = Date.now() - placeholdersVisibleStart;
+		debug(`Recording placeholder wait. Duration: ${duration}, Trigger: ${trigger}`);
+		analytics.timing.record( `placeholder-wait`, duration, trigger );
+		placeholdersVisibleStart = visibleCount === 0 ? null : Date.now();
+	}
+
+	// if we can see placeholders, placeholdersVisibleStart is falsy, start the clock
+	if ( visibleCount > 0 && !placeholdersVisibleStart ) {
+		placeholdersVisibleStart = Date.now(); // TODO: performance.now()?
+	}
+
+	// if there are placeholders hanging around, print some useful stats
+	if ( activePlaceholders.length > 0 ) {
+		debug("Active placeholders: "+activePlaceholders.length);
+		debug("Visible in viewport: "+visibleCount);
+	}
+}
+
+function isPlaceholder( node ) {
+	var className = node.className;
+	return className && className.indexOf && PLACEHOLDER_CLASSES.some( function( clazz ) { 
```

Comment:
`Array.prototype.some()` is actually now standard, in which I see you are using here, meaning we do not need to include **lodash**'s `some`